pygmt/projection.py

```diff
@@ -0,0 +1,590 @@
+>>> print(proj)
+A30/-20/60/8i
+"""
+
+from enum import Enum
+import attr
+
+
+class Supported(Enum):
+
+    """
+    The supported projections and their GMT code.
+    """
+
+    UNDEFINED = ""
+    LAMBERT_AZIMUTH_EQUAL_AREA = "A"  # DONE
+    ALBERS_CONIC_EQUAL_AREA = "B"  # DONE
+    CASSINI_CYLINDRICAL = "C"  # DONE
+    CYLINDRICAL_STEROGRAPHIC = "JCyl_stere/"  # includes `/` according to https://docs.generic-mapping-tools.org/latest/proj_codes.html  # DONE
+    EQUIDISTANT_CONIC = "JD"  # DONE
+    AZIMUTHAL_EQUIDISTANT = "E"  # DONE
+    AZIMUTHAL_GNOMIC = "F"  # DONE
+    AZIMUTHAL_ORTHOGRAPHIC = "G"  # DONE
+    GENERAL_PERSPECTIVE = "G"  # DONE
+    HAMMER_EQUAL_AREA = "H"
+    SINUSOIDAL_EQUAL_AREA = "I"
+    MILLER_CYLINDRICAL = "J"
+    ECKERT_IV_EQUAL_AREA = "Kf"
+    ECKERT_VI_EQUAL_AREA = "Ks"
+    LAMBERT_CONIC_CONFORMAL = "L"
+    MERCATOR_CYLINDRICAL = "M"  # DONE
+    ROBINSON = "N"
+    OBLIQUE_MERCATOR_1 = "Oa"
+    OBLIQUE_MERCATOR_2 = "Ob"
+    OBLIQUE_MERCATOR_3 = "Oc"
+    POLAR = "P"
+    POLYCONIC = "Poly"
+    EQUIDISTANT_CYLINDRICAL = "Q"
+    WINKEL_TRIPEL = "R"
+    GENERAL_STEREOGRAPHIC = "S"  # DONE
+    TRANSVERSE_MERCATOR = "T"
+    UNIVERSAL_TRANSVERSE_MERCATOR = "U"
+    VAN_DER_GRINTEN = "V"
+    MOLLWEIDE = "W"
+    LINEAR = "X"
+    CYLINDRICAL_EQUAL_AREA = "Y"  # DONE
+
+
+@attr.s()
+class _Projection:
+
+    """
+    Base class for all projections.
+    """
+
+    # private; we don't want the user to care or know about
+    _fmt: str = attr.ib(init=False, repr=False, default="{_code}")
+    _code: str = attr.ib(init=False, repr=False, default=Supported.UNDEFINED.value)
+
+    def __str__(self):
+        exclude = attr.fields(self.__class__)._fmt
+        kwargs = attr.asdict(self, filter=attr.filters.exclude(exclude))
+        return self._fmt.format(**kwargs)
+
+
+@attr.s(kw_only=True)
+class _Azimuthal(_Projection):
+
+    """
+    Base class for azimuthal projections.
+
+    Parameters
+    ----------
+    central_longitude : float
+        The longitude of the projection centre.
+    central_latitude : float
+        The latitude of the projection centre.
+    horizon : float
+        The max distance to the projection centre in degrees. Default is 90.
+    width : float
+        The figure width.
+    unit : str
+        The unit for the figure width in ``i`` for inch, ``c`` for centimetre.
+        Default is ``i``.
+    """
+
+    central_longitude: float = attr.ib()
+    central_latitude: float = attr.ib()
+    horizon: float = attr.ib(default=90)
+    width: float = attr.ib()
+    unit: str = attr.ib(default="i")
+    center: list = attr.ib(init=False)
+
+    # private; we don't want the user to care or know about
+    _fmt: str = attr.ib(
+        init=False,
+        repr=False,
+        default="{_code}{central_longitude}/{central_latitude}/{horizon}/{width}{unit}",
+    )
+    _code: str = attr.ib(init=False, repr=False, default=Supported.UNDEFINED.value)
```

Comment:
There is probably no need to define `_code` as "undefined" here. It's what has to be implemented by the base classes, right?

Reply:
As `_Projection`, `_Azimuthal`, `_Cylindrical` are considered private, I wouldn't expect the user to call them directly.  But they could and it would fail if `_code` is removed, as the `_fmt` attribute contains `{_code}`.  It is just an empty string placeholder, as such in order to avoid failure from a user calling it explicitly, both the attribute `_code` and the portion of the `str` containing `{_code}` could be removed.
Any thoughts on that approach?

Reply:
Actually, scratch that, I can't modify the formatter string, as it used by most of the subclasses.

Reply:
Woops again. I see what you mean.  I'll remove it, and maybe I should get to bed ;)